mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+// receive messages for, and slices for returning results.  Only non-nil slices
+// will be appended to.  Received messages are unsorted.
+type Received struct {
+	Sid [32]byte
+	Run uint32
+	KEs []*wire.MsgMixKeyExchange
+	CTs []*wire.MsgMixCiphertexts
+	SRs []*wire.MsgMixSlotReserve
+	DCs []*wire.MsgMixDCNet
+	CMs []*wire.MsgMixConfirm
+	FPs []*wire.MsgMixFactoredPoly
+	RSs []*wire.MsgMixSecrets
+}
+
+// Receive returns messages matching a session, run, and message type, waiting
+// until all described messages have been received, or earlier with the
+// messages received so far if the context is cancelled before this point.
+//
+// Receive only returns results for the session ID and run increment in the r
+// parameter.  If no such session or run has any messages currently accepted
+// in the mixpool, the method immediately errors.
+//
+// If any secrets messages are received for the described session and run, and
+// r.RSs is nil, Receive immediately returns ErrSecretsRevealed.  An
+// additional call to Receive with a non-nil RSs can be used to receive all of
+// the secrets after each peer publishes their own revealed secrets.
+func (p *Pool) Receive(ctx context.Context, expectedMessages int, r *Received) error {
+	sid := r.Sid
+	run := r.Run
+	var bc *broadcast
+	var rs *runstate
+	var err error
+
+	p.mtx.RLock()
+	ses, ok := p.sessions[sid]
+	if !ok {
+		p.mtx.RUnlock()
+		return fmt.Errorf("unknown session %x", sid[:])
+	}
+	bc = &ses.bc
+	if run >= uint32(len(ses.runs)) {
+		p.mtx.RUnlock()
+		return fmt.Errorf("unknown run %d", run)
+	}
+	rs = &ses.runs[run]
+
+Loop:
+	for {
+		// Pool is locked for reads.  Count if the total number of
+		// expected messages have been received.
+		received := 0
+		for hash := range rs.hashes {
+			msgtype := p.pool[hash].msgtype
+			switch {
+			case msgtype == msgtypeKE && r.KEs != nil:
+				received++
+			case msgtype == msgtypeCT && r.CTs != nil:
+				received++
+			case msgtype == msgtypeSR && r.SRs != nil:
+				received++
+			case msgtype == msgtypeDC && r.DCs != nil:
+				received++
+			case msgtype == msgtypeCM && r.CMs != nil:
+				received++
+			case msgtype == msgtypeFP && r.FPs != nil:
+				received++
+			case msgtype == msgtypeRS:
+				if r.RSs == nil {
+					// Since initial reporters of secrets
+					// need to take the blame for
+					// erroneous blame assignment if no
+					// issue was detected, we only trigger
+					// this for RS messages that do not
+					// reference any other previous RS.
+					rs := p.pool[hash].msg.(*wire.MsgMixSecrets)
+					prev := rs.PrevMsgs()
+					if len(prev) == 0 {
+						p.mtx.RUnlock()
+						return ErrSecretsRevealed
+					}
+				} else {
+					received++
+				}
+			}
+		}
+		if received >= expectedMessages {
+			break
+		}
+
+		// Unlock while waiting for the broadcast channel.
+		p.mtx.RUnlock()
+
+		select {
+		case <-ctx.Done():
+			// Set error to be returned, but still lock the pool
+			// and collect received messages.
+			err = ctx.Err()
```

Comment:
this could return a nil error.  all of the caller uses check the appended-to slice to see if enough messages were received (and have to check it in a multitude of other ways as well).